test/validation/api/shmem/shmem.c

```diff
@@ -212,6 +214,55 @@ void shmem_test_basic(void)
 	CU_ASSERT(0 == odp_shm_free(shm));
 }
 
+/*
+ * maximum size reservation
+ */
+static void shmem_test_max_reserve(void)
+{
+	odp_shm_capability_t capa;
+	odp_shm_t shm;
+	uint64_t size, align;
+	uint8_t *data;
+	uint64_t i;
+
+	memset(&capa, 0, sizeof(odp_shm_capability_t));
+	CU_ASSERT_FATAL(odp_shm_capability(&capa) == 0);
+
+	CU_ASSERT(capa.max_blocks > 0);
+
+	size  = capa.max_size;
+	align = capa.max_align;
+
+	/* Assuming that system has at least MAX_MEMORY_USED bytes available */
+	if (capa.max_size == 0 || capa.max_size > MAX_MEMORY_USED)
+		size = MAX_MEMORY_USED;
```

Comment:
Why do you want to limit the size in a test that named `shmem_test_max_reserve()`? If `capa.max_size == 0` then you have to pick a specific target, but if it's non-zero why wouldn't you want to try to reserve that much to see if the limit is true?

Reply:
Since actual amount of available memory typically depends on system load. SHM implementation may not have a limit (max_size==0), or limit may be due to address space (e.g. 40bit == 1TB). System might not have always the max amount (e.g. 1TB) available. I limit validation test to assume that at least 100MB should be always available.

Reply:
If the implementation doesn't have a specific predefined upper limit then it should return `capa.max_size == 0`. If it says it has a non-zero upper limit then if it's unable to provide that limit that's a failure. Otherwise what's the point of having a specified limit?

Reply:
Implementation may have a limitation of e.g. 4GB due to limit of using a 32 bit address space reservation, etc. It would be waste to reserve 4GB of system memory for every ODP instance, as  implementation could not guarantee 4GB otherwise, as other applications allocate memory as well. So, init  phase there could be 4.2GB available, but by the time ODP application starts calling shm_reserve() there would be less than 4GB left and some reserves would fail.

So, implementation may have large upper limit, which is not related to amount of available memory but e.g. due implementation of the address mapping (number of bits, hugepages, etc).

Reply:
The 0 value is intended to say that addressability / available memory is the only limit, so again in this case the implementation should return 0, not 4GB. If the implementation says 0 and the application tries to reserve something huge and that fails that's OK. The application needs to check RCs in any event. But what's the point of having a non-zero limit if there's no reasonable expectation that it means anything? At that point it's useless and might as well be ignored.

Reply:
In the example, system is 64 bit, has e.g. 32 GB of memory, ODP SHM implementation is limited e.g. by address space pre-reservation etc (e.g. 32 bits / 4GB). So, ODP implementation limits max SHM size to 4GB, 0 cannot be used as capability. A 4 GB reservation succeeds as long as system has memory free. When other applications or ODP instances have reserved all 32GB memory, yet another 4GB reservation will fail.

So, 0 says that there's no other limit than amount of currently free memory. E.g. odp-linux implementation lied by returning 0 and at the same time limiting max reservation to 512MB (due to fixed size address space allocation).